sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java

```diff
@@ -563,12 +563,29 @@ public static Unbound withSideInputs(
               DoFn.class.getSimpleName()));
     }
 
+    // To be removed when the features are complete and runners have their own adequate
+    // rejection logic
+    if (!signature.timerDeclarations().isEmpty()) {
+      throw new UnsupportedOperationException(
+          String.format("Found %s annotations on %s, but %s cannot yet be used with timers.",
+              DoFn.TimerId.class.getSimpleName(),
+              fn.getClass().getName(),
+              DoFn.class.getSimpleName()));
+    }
+
     // State is semantically incompatible with splitting
     if (!signature.stateDeclarations().isEmpty()) {
       throw new UnsupportedOperationException(
           String.format("%s is splittable and uses state, but these are not compatible",
               fn.getClass().getName()));
     }
+
+    // Timers are semantically incompatible with splitting (they are per-key)
+    if (!signature.timerDeclarations().isEmpty()) {
```

Comment:
This seems to duplicate the condition above, and also not test that it's splittable.